primitives/src/merkle_tree/merkle_tree_traits.rs

```diff
@@ -0,0 +1,158 @@
+    Ok(F, P),
+    /// The index is valid but we do not have the leaf in memory
+    NotInMemory,
+    /// The index is outside the occupied range in the tree
+    EmptyLeaf,
+}
+
+impl<F, P> LookupResult<F, P> {
+    /// Assert the lookup result is Ok.
+    pub fn expect_ok(self) -> Result<(F, P), PrimitivesError> {
+        match self {
+            LookupResult::Ok(x, proof) => Ok((x, proof)),
+            LookupResult::NotInMemory => Err(PrimitivesError::InternalError(
+                "Expected Ok, found NotInMemory".to_string(),
+            )),
+            LookupResult::EmptyLeaf => Err(PrimitivesError::InternalError(
+                "Expected Ok, found EmptyLeaf".to_string(),
+            )),
+        }
+    }
+}
+
+/// Merkle tree element type
+pub trait ElementType<F: Field>:
+    Default
+    + Ord
+    + Clone
+    + Copy
+    + Debug
+    + CanonicalSerialize
+    + CanonicalDeserialize
+    + Eq
+    + PartialEq
+    + AsRef<[F]>
+{
+    /// Convert the element into a slice of field elements
+    /// TODO: is it the same as `AsRef<[F]>`?
+    fn as_slice_ref(&self) -> &[F];
+}
+
+/// Merkle tree hash function
+pub trait Hasher<F: Field> {
+    /// Digest a list of values
+    fn digest(data: &[F]) -> F;
+}
+
+/// Basic functionalities for a merkle tree implementation
+pub trait MerkleTree<F: Field>: Sized {
+    /// Merkle tree element type
+    type ElementType: ElementType<F>;
+    /// Hash algorithm used in merkle tree
+    type Hasher: Hasher<F>;
+    /// Leaf arity
+    type LeafArity: Unsigned;
+    /// Non-leaf arity
+    type TreeArity: Unsigned;
+    /// Exsistential proof
+    type Proof;
+    /// Batch proof
+    type BatchProof;
+
+    /// Construct a new merkle tree with given height from a data slice
+    fn build(
+        height: usize,
+        data: impl Iterator<Item = Self::ElementType>,
+    ) -> Result<Self, PrimitivesError>;
+
+    /// Return the height of this merkle tree
+    fn height(&self) -> usize;
+    /// Return the maximum allowed number leaves
+    fn capacity(&self) -> u64;
+    /// Return the current number of leaves
+    fn num_leaves(&self) -> u64;
+
+    /// Return the current root value
+    fn value(&self) -> F;
+
+    /// Returns the leaf value given a position
+    /// * `pos` - zero-based index of the leaf in the tree
+    /// * `returns` - Leaf value at the position. LookupResult::EmptyLeaf if the
+    ///   leaf position is empty or invalid, LookupResult::NotInMemory if the
+    ///   leaf position has been forgotten.
+    fn lookup(&self, pos: u64) -> LookupResult<(), Self::Proof>;
+
+    /// Verify an element is a leaf of a Merkle tree given the proof
+    /// * `pos` - zero-based index of the leaf in the tree
+    /// * `proof` - a merkle tree proof
+    /// * `returns` - Err() if something is wrong with this proof, Ok(result)
+    ///   otherwise
+    fn verify(&self, pos: u64, proof: impl Borrow<Self::Proof>) -> Result<bool, PrimitivesError>;
```

Comment:
What does `true`/`false` indicate in the return value?

Reply:
Then what does `Err` indicate? I thought `Err` would mean the proof is not valid and `Ok` would mean it was. I think at least this function needs better documentation on how to interpret the result.

Reply:
`Err` indicates that the proof is not well structured (E.g. not for this given MT).
Sure I'll try to document it better.

Reply:
Oh, I see. Makes sense now